### ISSUE TITLE
Add trunk to VLAN

### ISSUE THREAD

**Issue by saro** [Broadcom-Switch/OpenNSL#14]
Hello,
sorry if the question looks simple but I'm a bit lost in the documentation. I would like to add a trunk to specific vlan. Is there any example for this use case?

**Comment by simonknee** [comment]
Take a look at the example_trunk.c application.  It shows how to add/delete a port to/from the trunk in the default VLAN.

**Comment by saro** [comment]
Thank for the replay @simonknee 
I have a second question, is it possible to restrict the trunk for a specific set of VLANs?

**Comment by simonknee** [comment]
First, just to correct my last answer: you were asking about adding a trunk to a VLAN but I answered about adding ports to a trunk.  Since a trunk is a logical interface, there are no APIs to add a trunk to a VLAN. Instead, the individual trunk member ports need to be added to the VLAN.  This needs to be handled by the application.  As and when trunk port members are added or removed, the VLAN bitmap needs to be updated.

It is also the application which controls which VLANs a trunk (or its port numbers) is a member of, since it controls adding/removing them.

**Comment by saro** [comment]
So, If my understanding is correct, given two VLAN and a Trunk with the same ports it's not possible to restrict the trunk to load balance only one of the two VLAN.
Is this correct?

**Comment by simonknee** [comment]
Yes, that is correct.  The reason is that the hashing (PSC - Port Selection Criteria) is configurable at a trunk port level, but not at a VLAN level.  See opennsl_trunk_psc_set() for how to configure hashing on a trunk port.

**Comment by saro** [comment]
Thanks again for the clarification.
I have just a last question regarding the trunking and hashing. In the `example_trunking.c` I can see that some hashing options are enabled in the function `example_trunk_hash_controls_set`.
Are these enough to use all the available PSC options? or are additional steps necessary?

**Comment by simonknee** [comment]
No problem Saro.


The following supported PSC options are available in include/opennsl/trunk.h file.
 
#define OPENNSL_TRUNK_PSC_SRCDSTMAC 3          /**< Source+dest MAC address. */
#define OPENNSL_TRUNK_PSC_DSTIP     5          /**< Destination IP address. */
#define OPENNSL_TRUNK_PSC_SRCDSTIP  6          /**< Source+dest IP address. */
#define OPENNSL_TRUNK_PSC_PORTFLOW  9          /**< Enhanced hashing. */
 
When a trunk is created, the default PSC (OPENNSL_TRUNK_PSC_SRCDSTMAC) is used, where hashing is based on the source and destination MAC address. The sample application provides the user an option (7) to change the PSC, and lets you change all PSC options.

**Comment by saro** [comment]
Thanks the explanation.
What is not totally clear to me is the role of the following switch options:

OPENNSL_HASH_CONTROL_TRUNK_NUC_DST;
OPENNSL_HASH_CONTROL_TRUNK_NUC_SRC;
OPENNSL_HASH_CONTROL_TRUNK_UC_SRCPORT;
OPENNSL_HASH_CONTROL_MULTIPATH_DIP;
OPENNSL_HASH_CONTROL_MULTIPATH_L4PORTS;

Are these always necessary? Do they impact all the PSC options?
and are there other options that can be used to change the load balancing behaviour?

**Comment by sivamukka** [comment]
The hash controls OPENNSL_HASH_CONTROL_xxx are configurable at a Switch level and OPENNSL_TRUNK_PSC_xxx are configurable at a trunk port level. It's recommended to configure the Switch level hash controls (OPENNSL_HASH_CONTROL_XXX) also, as these
=> Help in hashing of non-unicast traffic 
=> Takes TCP/UDP port parameters for hash criteria
=> Used for ECMP traffic hashing
